### PR TITLE
`BinaryAUROC` support for Masked Labels

### LINKED ISSUE #3096: Provide BinaryAUROC support for Masked/Sparse Labels
## 🚀 Feature

Currently the `BinaryAUROC` metric update step expects a square matrix at each iteration. In a multi-output training regiment involving masked outputs, the current AUROC metrics class fails to offer support.

This feature proposes to provide an additional metric, `MaskedBinaryAUROC`, to support calculating label-specific and aggregate AUROC when each batch step leverages sparse/nested/masked labels.

### Motivation

I had to create a custom class for this functionality for a professional project, and I'd like to contribute it. Also, my company would like to become involved in the open source community by promoting code when possible.

### Pitch

Create a `MaskedBinaryAUROC` metric with similar functionality to `AUROC`, but it will also take a mask at each iteration and only consider unmasked values in the final calculations.

The implementation is straightforward:

0. preds/targets/mask states with List defaults
1. `.update(preds, targets, mask)` by appending to the list
2. `.compute()` will iterate through each column to calculate a per-label value by leveraging `torch.metrics.functional.binary_auroc` after applying the mask
3. return mean of all column-wise auroc values

### Alternatives

- `SparseBinaryAUROC`: support a sparse tensor representation of the output labels instead of a dense output and a mask.
- Extend functionality of `BinaryAUROC` instead of making a separate class

### Additional context

There are additional related extensions of the `AUROC`, `MulticlassAUROC`, and `MultilabelAUROC` classes.

### PR DESCRIPTION
**Pull request by VijayVignesh1**
## What does this PR do?

Fixes #3096

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>
<br>
Added MaskedBinaryAUROC class implementation which updates the preds and target list with the corresponding mask. The other methods (compute, plot) remain the same. <br>
Added all the relavant testcases. <br><br>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/unittests/classification/test_auroc.py [hunk 1]
```diff
@@ -140,6 +148,157 @@ def test_binary_auroc_threshold_arg(self, inputs, threshold_fn):
...
+            preds.requires_grad = True
+            out = metric(preds[0, :2], target[0, :2], mask[0, :2] if mask is not None else None)
+
+            # Check if requires_grad matches is_differentiable attribute
+            _assert_requires_grad(metric, out)
+
+            if metric.is_differentiable and metric_functional is not None:
+                # check for numerical correctness
+                assert torch.autograd.gradcheck(
+                    partial(metric_functional, **metric_args), (preds[0, :2].double(), target[0, :2])
```
[the remark this answers is not in the document]
Comment (VijayVignesh1) [reply tests/unittests/classification/test_auroc.py:207]: The functional version is still using binary_auroc which doesn't have the mask parameter. So testing the metric_function with mask parameter is irrelevant.

File: src/torchmetrics/classification/auroc.py [hunk 2]
```diff
@@ -167,6 +168,124 @@ def plot(  # type: ignore[override]
...
+        tensor(0.5000)
+        >>> b_auroc = MaskedBinaryAUROC(thresholds=5)
+        >>> b_auroc(preds, target, mask)
+        tensor(0.5000)
+
+    """
+
+    def update(self, preds: Tensor, target: Tensor, mask: Optional[Tensor] = None) -> None:
+        """Update the state with the new data."""
+        if mask is not None:
```
Comment (justusschock) [inline src/torchmetrics/classification/auroc.py]: in masked auroc, mask cannot be none, let's update the typing and remove the check

File: src/torchmetrics/classification/auroc.py [hunk 3]
```diff
@@ -167,6 +168,124 @@ def plot(  # type: ignore[override]
         return self._plot(val, ax)
 
 
+class MaskedBinaryAUROC(BinaryAUROC):
```
Comment (justusschock) [inline src/torchmetrics/classification/auroc.py:171]: do we need a functional version of this as well?

File: src/torchmetrics/classification/auroc.py (hunk 2 above)
Comment (VijayVignesh1) [reply src/torchmetrics/classification/auroc.py]: Sure, I'll make the mask mandatory and remove the check.

File: src/torchmetrics/classification/auroc.py (hunk 3 above)
Comment (VijayVignesh1) [reply src/torchmetrics/classification/auroc.py:171]: We don't need the functional version since I'm basically reusing the binaryauroc function. The only change will be in the way we update the preds and target. The rest remains the same.
